File: android/domain/src/main/kotlin/com/jarvismobile/domain/tooling/SkillRegistry.kt
package com.jarvismobile.domain.tooling

import com.jarvismobile.domain.entity.SkillCategory
import com.jarvismobile.domain.entity.SkillDefinition

/**
 * In-memory catalogue of every registered skill. Populated at app/backend startup by each
 * skill package (see SKILLS.md "Authoring a new skill"). Drives both the LLM tool-calling
 * definitions (AI_ARCHITECTURE.md) and the live "What can you do?" screen — never hand-authored
 * per screen.
 */
class SkillRegistry {
    private val skills = mutableMapOf<String, SkillDefinition>()

    fun register(skill: SkillDefinition) {
        require(skill.id !in skills) { "Duplicate skill id: '${skill.id}' is already registered" }
        skills[skill.id] = skill
    }

    fun find(id: String): SkillDefinition? = skills[id]

    fun all(): List<SkillDefinition> = skills.values.toList()

    fun byCategory(category: SkillCategory): List<SkillDefinition> =
        skills.values.filter { it.category == category }
}

File: backend/src/tooling/skillRegistry.ts
import type { SkillCategory, SkillDefinition } from "../domain/types.js";
import { assertValidSkillId } from "../domain/types.js";

/**
 * In-memory catalogue of every registered skill, mirroring android/domain/tooling/SkillRegistry.kt.
 * Drives GET /api/v1/skills (the live "What can you do?" catalogue) and the Orchestrator's
 * tool-calling loop — see SKILLS.md.
 */
export class SkillRegistry {
  private readonly skills = new Map<string, SkillDefinition>();

  register(skill: SkillDefinition): void {
    assertValidSkillId(skill.id);
    if (this.skills.has(skill.id)) {
      throw new Error(`Duplicate skill id: '${skill.id}' is already registered`);
    }
    this.skills.set(skill.id, skill);
  }

  find(id: string): SkillDefinition | undefined {
    return this.skills.get(id);
  }

  all(): SkillDefinition[] {
    return [...this.skills.values()];
  }

  byCategory(category: SkillCategory): SkillDefinition[] {
    return this.all().filter((skill) => skill.category === category);
  }
}
